ts/screens/authentication/IdpLoginScreen.tsx

```diff
@@ -94,34 +117,109 @@ const onNavigationStateChange = (
  * The IDP page is opened in a WebView
  */
 class IdpLoginScreen extends React.Component<Props, State> {
+  private loginTrace: LoginTrace;
+
   constructor(props: Props) {
     super(props);
-
+    const nowMs = new Date().getTime();
+    this.loginTrace = {
+      deltaMillis: nowMs as Millisecond
+    };
     this.state = {
       requestState: pot.noneLoading
     };
   }
 
-  private handleOnError = (): void =>
+  private updateLoginTrace(delta: Millisecond, url?: string): void {
+    // tslint:disable-next-line: no-object-mutation
+    this.loginTrace = {
+      deltaMillis: delta,
+      requestUrl: url || this.loginTrace.requestUrl
+    };
+  }
+
+  private getIdpId = (): string =>
+    this.props.loggedOutWithIdpAuth
+      ? this.props.loggedOutWithIdpAuth.idp.entityID
+      : "idpID n/a";
+
+  public componentDidMount() {
+    this.props.dispatchAction(
+      idpLoginSession({
+        idpId: this.getIdpId(),
+        started: true
+      })
+    );
+  }
+
+  private handleOnError = (): void => {
+    const now = new Date().getTime();
+    const deltaDuration = Math.round(
+      (now - this.loginTrace.deltaMillis) / 1000
+    );
+    this.props.dispatchAction(
+      idpLoginRequestError({
+        idpId: this.getIdpId(),
+        durationMillis: deltaDuration as Millisecond
+      })
+    );
     this.setState({
       requestState: pot.noneError("error")
     });
+    this.updateLoginTrace(now as Millisecond);
+  };
 
   private goBack = this.props.navigation.goBack;
 
-  private setRequestStateToLoading = () =>
+  private setRequestStateToLoading = (): void =>
     this.setState({ requestState: pot.noneLoading });
 
   private handleNavigationStateChange = (event: NavState): void => {
+    const now = new Date().getTime();
+    if (event.url && event.url !== this.loginTrace.requestUrl) {
+      const deltaDuration = Math.round(
+        (now - this.loginTrace.deltaMillis) / 1000
+      );
+      this.props.dispatchAction(
+        idpLoginUrlChanged({
+          idpId: this.getIdpId(),
+          url: event.url.split("?")[0],
+          durationMillis: deltaDuration as Millisecond
+        })
+      );
+    }
+    this.updateLoginTrace(now as Millisecond, event.url);
     this.setState({
       requestState: event.loading
-        ? pot.someLoading("loading")
-        : pot.some("loading complete")
+        ? pot.someLoading("LOADING")
```

Comment:
what's the meaning of `pot.someLoading("LOADING")`, "some loading" means I have a value and I'm loading a new one

Reply:
aka Updating

Reply:
yeah updating *from* remote source - there's actually also a `someUpdating(curr, next)` state for updating the remote state with a new (`next`) value where `prev` is the previous one (i.e. when saving an updated value to the remote state, and in case of error you want to rollback)